contrib/pg_tde/src/access/pg_tde_xlog_smgr.c

```diff
@@ -299,26 +336,64 @@ static ssize_t
 tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset,
 					   TimeLineID tli, XLogSegNo segno, int segSize)
 {
+	bool		lastKeyUsable;
+	bool		afterLastKey;
+#ifdef FRONTEND
+	bool		crashRecovery = false;
+#else
+	bool		crashRecovery = GetRecoveryState() == RECOVERY_STATE_CRASH;
+#endif
+
+	WalLocation loc = {.tli = tli};
+	WalLocation last_key_loc;
+
+	XLogSegNoOffsetToRecPtr(segno, offset, segSize, loc.lsn);
+
 	/*
 	 * Set the last (most recent) key's start LSN if not set.
 	 *
 	 * This func called with WALWriteLock held, so no need in any extra sync.
 	 */
-	if (EncryptionKey.type != WAL_KEY_TYPE_INVALID && TDEXLogGetEncKeyLsn() == 0)
-	{
-		WalLocation loc = {.tli = tli};
 
-		XLogSegNoOffsetToRecPtr(segno, offset, segSize, loc.lsn);
+	last_key_loc.lsn = TDEXLogGetEncKeyLsn();
+	pg_read_barrier();
+	last_key_loc.tli = TDEXLogGetEncKeyTli();
+
+	lastKeyUsable = (TDEXLogGetEncKeyLsn() != 0);
+	afterLastKey = wal_location_cmp(last_key_loc, loc) <= 0;
 
-		pg_tde_wal_last_key_set_location(loc);
-		EncryptionKey.wal_start = loc;
-		TDEXLogSetEncKeyLocation(EncryptionKey.wal_start);
+	if (EncryptionKey.type != WAL_KEY_TYPE_INVALID && !lastKeyUsable)
+	{
+		WALKeyCacheRec *last_key = pg_tde_get_last_wal_key();
+
+		if (!crashRecovery || EncryptionKey.type == WAL_KEY_TYPE_UNENCRYPTED)
```

Comment:
I would want a comment here explain why the unencrypted case does not need to care, because on top of my head I do not udnerstand why it would be special.

Reply:
It has to be special, we'll need more conditions around that - but that's not covered by the testcase, and I still have to figure out exactly what, that's why I left it as is and left a todo comment related to it.

Reply:
The more I think the less I get why it needs to be special. Don't we need to make sure old entries which are encryted stay encrypted with the same keys?